Structured output logits processors become misaligned when generation batches are extended

I'm reopening this issue because the previous one (#1555) got closed with a PR that didn't fix the issue I was reporting. I've added another testing script which loads an actual model if you prefer that to reproduce the issue

## Description

`GenerationBatch.extend()` can misalign per-request logits processors when structured and unstructured requests are admitted into a continuous generation batch in separate waves.

Batches without active processors use an empty `logits_processors` list as a compact representation. If multiple plain batches are merged and a structured request is appended afterward, the structured request's processor is appended without placeholders for the existing rows.

For example, three UIDs may end up with only one processor entry:

```text
uids: [0, 1, 2]
logits_processors: [[structured_processor]]
```

The processor belongs to UID 2, but occupies row 0. During decoding, the structured-output grammar can therefore be applied to an unrelated plain request. Subsequent rows may also encounter an out-of-range processor index. I initially encountered this through concurrent structured and unstructured requests using the mlx-vlm continuous server, but the problem reproduces directly with GenerationBatch.

## Reproduction

Run the following script from an mlx-vlm checkout:

```python
import mlx.core as mx

from mlx_vlm.generate.ar import GenerationBatch

sampler = lambda logprobs: mx.argmax(logprobs, axis=-1)
stop_criteria = lambda token: False

def make_batch(uid, processor=None):
    return GenerationBatch(
        model=object(),
        uids=[uid],
        inputs=mx.array([uid + 1], dtype=mx.int32),
        prompt_cache=[],
        sampler=sampler,
        stop_criteria=stop_criteria,
        max_tokens=[2],
        token_context=[[30]] if processor is not None else None,
        logits_processors=[[processor]] if processor is not None else None,
    )

first_plain = make_batch(0)
second_plain = make_batch(1)

structured_processor = lambda tokens, logits: logits
structured = make_batch(2, structured_processor)

first_plain.extend(second_plain)
print("After plain + plain:", first_plain.logits_processors)

first_plain.extend(structured)
print("UIDs:", first_plain.uids)
print("Processor slots:", first_plain.logits_processors)

assert len(first_plain.logits_processors) == len(first_plain.uids)
assert first_plain.logits_processors == [
    None,
    None,
    [structured_processor],
]
```

## Actual behavior

The final processor list contains only the structured processor:

```
After plain + plain: []
UIDs: [0, 1, 2]
Processor slots: [[<function ...>]]
```

The first assertion fails:

`AssertionError: ([0, 1, 2], [[<function ...>]])`

The processor is incorrectly associated with row 0, although it belongs to UID 2.
The reverse admission order is also unsafe: appending a plain request to an active structured batch leaves the processor list shorter than the UID list.

## Expected behavior

When any request in the merged batch has an active processor, every UID should have a corresponding processor slot:

```
UIDs: [0, 1, 2]
Processor slots: [None, None, [structured_processor]]
```

Plain rows should retain None, and the structured processor should remain aligned with its originating request.

## Repro with a model

Here is another reproduction script that loads an actual Qwen model if you prefer that (just update the path accordingly):

```python
import concurrent.futures
import json
import threading

from mlx_vlm.generate.ar import GenerationBatch
from mlx_vlm.prompt_utils import apply_chat_template
from mlx_vlm.server import GenerationArguments, ResponseGenerator
from mlx_vlm.structured import build_json_schema_logits_processor

MODEL_PATH = (
    "/Users/yoan/.lmstudio/models/lmstudio-community/"
    "Qwen3.5-9B-MLX-4bit"
)

plain_batch_ready = threading.Event()
misalignment_seen = threading.Event()
original_extend = GenerationBatch.extend

def traced_extend(self, other):
    result = original_extend(self, other)
    processors = self.logits_processors
    active_processors = bool(processors and any(processors))
    print(
        "GenerationBatch.extend:",
        {
            "uids": list(self.uids),
            "processor_slots": len(processors),
            "active_slots": [
                index for index, value in enumerate(processors) if value
            ],
        },
        flush=True,
    )
    if len(self.uids) >= 2 and not active_processors:
        plain_batch_ready.set()
    if active_processors and len(processors) != len(self.uids):
        misalignment_seen.set()
        print(
            "MISALIGNED PROCESSORS:",
            {"uids": list(self.uids), "processors": processors},
            flush=True,
        )
    return result

GenerationBatch.extend = traced_extend

def consume(name, iterator):
    text = []
    error = None
    try:
        for token in iterator:
            piece = getattr(token, "text", "") or ""
            text.append(piece)
    except Exception as exc:
        error = f"{type(exc).__name__}: {exc}"
    return {"name": name, "text": "".join(text)[:240], "error": error}


generator = ResponseGenerator(model_path=MODEL_PATH)
try:
    _, processor, config = generator.wait_until_ready(timeout=300)
    tokenizer = (
        processor.tokenizer if hasattr(processor, "tokenizer") else processor
    )

    def format_prompt(text):
        return apply_chat_template(
            processor,
            config,
            text,
            num_images=0,
            enable_thinking=False,
        )

    plain_args = GenerationArguments(
        max_tokens=512,
        temperature=0,
        enable_thinking=False,
    )
    schema = {
        "type": "object",
        "additionalProperties": False,
        "properties": {
            "answer": {"type": "string"},
        },
        "required": ["answer"],
    }
    schema_processor = build_json_schema_logits_processor(tokenizer, schema)
    structured_args = GenerationArguments(
        max_tokens=64,
        temperature=0,
        enable_thinking=False,
        logits_processors=[schema_processor],
    )

    _, plain_one = generator.generate(
        prompt=format_prompt(
            "Write a long, detailed explanation of continuous batching."
        ),
        args=plain_args,
    )
    _, plain_two = generator.generate(
        prompt=format_prompt(
            "Write a long, detailed explanation of structured generation."
        ),
        args=plain_args,
    )

    if not plain_batch_ready.wait(timeout=180):
        raise RuntimeError("Two plain requests did not merge before timeout")

    _, structured = generator.generate(
        prompt=format_prompt("Return a short greeting in the requested schema."),
        args=structured_args,
    )

    with concurrent.futures.ThreadPoolExecutor(max_workers=3) as pool:
        futures = [
            pool.submit(consume, "plain_one", plain_one),
            pool.submit(consume, "plain_two", plain_two),
            pool.submit(consume, "structured", structured),
        ]
        results = [future.result(timeout=300) for future in futures]

    print(json.dumps({"results": results}, indent=2), flush=True)
    print(f"misalignment_seen={misalignment_seen.is_set()}", flush=True)
    if not misalignment_seen.is_set():
        raise AssertionError("Expected processor misalignment was not observed")
finally:
    generator.stop_and_join()
```

Output:

```bash
Loading model from: /Users/yoan/.lmstudio/models/lmstudio-community/Qwen3.5-9B-MLX-4bit
Model and processor loaded successfully.
GenerationBatch.extend: {'uids': [0, 1], 'processor_slots': 0, 'active_slots': []}
GenerationBatch.extend: {'uids': [0, 1, 2], 'processor_slots': 1, 'active_slots': [0]}
MISALIGNED PROCESSORS: {'uids': [0, 1, 2], 'processors': [[<mlx_vlm.structured.LLGuidanceLogitsProcessor object at 0x3ffeb82f0>]]}
Warning: Parser Error: token "atching" doesn't satisfy the grammar; forced bytes: got '{'; applying 'a'; stopping
[...]
ValueError: LLGuidance matcher error: Parser Error: token "atching" doesn't satisfy the grammar; forced bytes: got '{'; applying 'a'
<state>
Tokens: ⟦atching⟧
1 tokens, 0 bytes; grm_prefix: ""
Flags:
Parser: {
  "compute_time_us": 0,
  "rows": 7,
  "cached_rows": 438,
  "all_items": 7,
  "lexer_cost": 3214,
  "slices_applied": 0,
  "trie_nodes_walked": 0,
  "definitive_bytes": 12,
  "lexer_ops": 0,
  "num_lex_errors": 0,
  "num_lexemes": 0
}
Stop: ParserTooComplex
Error: Parser Error: token "atching" doesn't satisfy the grammar; forced bytes: got '{'; applying 'a'
</state><grammar>
{"grammars":[{"name":"json_schema","json_schema":{"type":"object","additionalProperties":false,"properties":{"answer":{"type":"string"}},"required":["answer"],"x-guidance":{"item_separator":", ","key_separator":": ","whitespace_flexible":false,"whitespace_pattern":"","coerce_one_of":false,"lenient":false,"json_allowed_escapes":null}},"lark_grammar":null}],"max_tokens":null}
</grammar>
{
  "results": [
    {
      "name": "plain_one",
      "text": "# Continuous B",
      "error": "ValueError: LLGuidance matcher error: Parser Error: token \"atching\" doesn't satisfy the grammar; forced bytes: got '{'; applying 'a'\n<state>\nTokens: \u27e6atching\u27e7\n1 tokens, 0 bytes; grm_prefix: \"\"\nFlags:\nParser: {\n  \"compute_time_us\": 0,\n  \"rows\": 7,\n  \"cached_rows\": 438,\n  \"all_items\": 7,\n  \"lexer_cost\": 3214,\n  \"slices_applied\": 0,\n  \"trie_nodes_walked\": 0,\n  \"definitive_bytes\": 12,\n  \"lexer_ops\": 0,\n  \"num_lex_errors\": 0,\n  \"num_lexemes\": 0\n}\nStop: ParserTooComplex\nError: Parser Error: token \"atching\" doesn't satisfy the grammar; forced bytes: got '{'; applying 'a'\n</state><grammar>\n{\"grammars\":[{\"name\":\"json_schema\",\"json_schema\":{\"type\":\"object\",\"additionalProperties\":false,\"properties\":{\"answer\":{\"type\":\"string\"}},\"required\":[\"answer\"],\"x-guidance\":{\"item_separator\":\", \",\"key_separator\":\": \",\"whitespace_flexible\":false,\"whitespace_pattern\":\"\",\"coerce_one_of\":false,\"lenient\":false,\"json_allowed_escapes\":null}},\"lark_grammar\":null}],\"max_tokens\":null}\n</grammar>"
    },
    {
      "name": "plain_two",
      "text": "# Struct",
      "error": "ValueError: LLGuidance matcher error: Parser Error: token \"atching\" doesn't satisfy the grammar; forced bytes: got '{'; applying 'a'\n<state>\nTokens: \u27e6atching\u27e7\n1 tokens, 0 bytes; grm_prefix: \"\"\nFlags:\nParser: {\n  \"compute_time_us\": 0,\n  \"rows\": 7,\n  \"cached_rows\": 438,\n  \"all_items\": 7,\n  \"lexer_cost\": 3214,\n  \"slices_applied\": 0,\n  \"trie_nodes_walked\": 0,\n  \"definitive_bytes\": 12,\n  \"lexer_ops\": 0,\n  \"num_lex_errors\": 0,\n  \"num_lexemes\": 0\n}\nStop: ParserTooComplex\nError: Parser Error: token \"atching\" doesn't satisfy the grammar; forced bytes: got '{'; applying 'a'\n</state><grammar>\n{\"grammars\":[{\"name\":\"json_schema\",\"json_schema\":{\"type\":\"object\",\"additionalProperties\":false,\"properties\":{\"answer\":{\"type\":\"string\"}},\"required\":[\"answer\"],\"x-guidance\":{\"item_separator\":\", \",\"key_separator\":\": \",\"whitespace_flexible\":false,\"whitespace_pattern\":\"\",\"coerce_one_of\":false,\"lenient\":false,\"json_allowed_escapes\":null}},\"lark_grammar\":null}],\"max_tokens\":null}\n</grammar>"
    },
    {
      "name": "structured",
      "text": "",
      "error": "ValueError: LLGuidance matcher error: Parser Error: token \"atching\" doesn't satisfy the grammar; forced bytes: got '{'; applying 'a'\n<state>\nTokens: \u27e6atching\u27e7\n1 tokens, 0 bytes; grm_prefix: \"\"\nFlags:\nParser: {\n  \"compute_time_us\": 0,\n  \"rows\": 7,\n  \"cached_rows\": 438,\n  \"all_items\": 7,\n  \"lexer_cost\": 3214,\n  \"slices_applied\": 0,\n  \"trie_nodes_walked\": 0,\n  \"definitive_bytes\": 12,\n  \"lexer_ops\": 0,\n  \"num_lex_errors\": 0,\n  \"num_lexemes\": 0\n}\nStop: ParserTooComplex\nError: Parser Error: token \"atching\" doesn't satisfy the grammar; forced bytes: got '{'; applying 'a'\n</state><grammar>\n{\"grammars\":[{\"name\":\"json_schema\",\"json_schema\":{\"type\":\"object\",\"additionalProperties\":false,\"properties\":{\"answer\":{\"type\":\"string\"}},\"required\":[\"answer\"],\"x-guidance\":{\"item_separator\":\", \",\"key_separator\":\": \",\"whitespace_flexible\":false,\"whitespace_pattern\":\"\",\"coerce_one_of\":false,\"lenient\":false,\"json_allowed_escapes\":null}},\"lark_grammar\":null}],\"max_tokens\":null}\n</grammar>"
    }
  ]
}
misalignment_seen=True
```

Running this same script with the fix from the PR attached to this issue:

```bash
Loading model from: /Users/yoan/.lmstudio/models/lmstudio-community/Qwen3.5-9B-MLX-4bit
Model and processor loaded successfully.
GenerationBatch.extend: {'uids': [0, 1], 'processor_slots': 0, 'active_slots': []}
GenerationBatch.extend: {'uids': [0, 1, 2], 'processor_slots': 3, 'active_slots': [2]}
{
  "results": [
    {
      "name": "plain_one",
      "text": "# Continuous Batching: A Deep Dive into Modern Data Processing Architectures\n\nIn the rapidly evolving landscape of data engineering, machine learning operations (MLOps), and high-performance computing, the paradigm of **Continuous Batching*",
      "error": null
    },
    {
      "name": "plain_two",
      "text": "# Structured Generation: The Architecture of Order in the Age of AI\n\nIn the rapidly evolving landscape of artificial intelligence, particularly within Large Language Models (LLMs), **Structured Generation** represents a paradigm shift from ",
      "error": null
    },
    {
      "name": "structured",
      "text": "{\"answer\": \"Hello! How can I help you today?\"}",
      "error": null
    }
  ]
}
misalignment_seen=False
```

## Impact

Under continuous batching, concurrent structured and unstructured requests can affect one another:

- JSON grammar being applied to an unrelated plain request
- Structured parser errors caused by tokens from the wrong request
- Invalid or truncated output
- Generation failure from processor-list indexing

Requests created together with an already aligned processor list are unaffected. The issue occurs when active generation batches are extended with requests that have different processor requirements.

--- a/mlx_vlm/generate/ar.py
+++ b/mlx_vlm/generate/ar.py
@@ -921,6 +921,15 @@ class GenerationBatch:
     def cache_states(self):
         return [c.state for c in self.prompt_cache if hasattr(c, "state")]
 
+    def _ensure_logits_processor_slots(self, *, force: bool = False):
+        if not (force or (self.logits_processors and any(self.logits_processors))):
+            return
+        if len(self.logits_processors) < len(self.uids):
+            missing = len(self.uids) - len(self.logits_processors)
+            self.logits_processors.extend([None] * missing)
+        elif len(self.logits_processors) > len(self.uids):
+            self.logits_processors = self.logits_processors[: len(self.uids)]
+
     def _ensure_token_context(self, *, force: bool = False):
         if not (force or (self.logits_processors and any(self.logits_processors))):
             if not self.logits_processors:
@@ -1102,6 +1111,8 @@ class GenerationBatch:
         self_has_processors = self.logits_processors and any(self.logits_processors)
         other_has_processors = other.logits_processors and any(other.logits_processors)
         if self_has_processors or other_has_processors:
+            self._ensure_logits_processor_slots(force=bool(other_has_processors))
+            other._ensure_logits_processor_slots(force=bool(self_has_processors))
             self._ensure_token_context(force=bool(other_has_processors))
             other._ensure_token_context(force=bool(self_has_processors))
         else:
@@ -1117,6 +1128,7 @@ class GenerationBatch:
         self.token_context.extend(other.token_context)
         self.logits_processors.extend(other.logits_processors)
         self.thinking_budget_criteria.extend(other.thinking_budget_criteria)
+        self._ensure_logits_processor_slots()
         self._ensure_token_context()
 
         if self._current_tokens is None:

--- a/mlx_vlm/tests/test_generate.py
+++ b/mlx_vlm/tests/test_generate.py
@@ -826,6 +826,41 @@ class TestBatchGenerator:
         assert [r.token for r in first] == [5, 6, 7]
         assert seen_contexts == [[30, 7]]
 
+    def test_generation_batch_extend_expands_compact_processor_state(self):
+        sampler = lambda logprobs: mx.argmax(logprobs, axis=-1)
+        stop_criteria = lambda token: False
+
+        def make_batch(uid, processor=None):
+            return GenerationBatch(
+                model=object(),
+                uids=[uid],
+                inputs=mx.array([uid + 1], dtype=mx.int32),
+                prompt_cache=[],
+                sampler=sampler,
+                stop_criteria=stop_criteria,
+                max_tokens=[2],
+                token_context=[[30]] if processor is not None else None,
+                logits_processors=[[processor]] if processor is not None else None,
+            )
+
+        first_plain = make_batch(0)
+        second_plain = make_batch(1)
+        structured_processor = lambda tokens, logits: logits
+        structured = make_batch(2, structured_processor)
+
+        first_plain.extend(second_plain)
+        assert first_plain.logits_processors == []
+
+        first_plain.extend(structured)
+
+        assert first_plain.uids == [0, 1, 2]
+        assert first_plain.token_context == [[], [], [30]]
+        assert first_plain.logits_processors == [
+            None,
+            None,
+            [structured_processor],
+        ]
+
     def test_generation_batch_extend_discards_inactive_stale_processor_state(self):
         sampler = lambda logprobs: mx.argmax(logprobs, axis=-1)
         stop_criteria = lambda token: False

